### PR TITLE
Staging to master: large refactor, LighGCN and NextItNet

### PR DESCRIPTION
**Pull request by miguelgfierro**
### Checklist:

- [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [ ] This PR is being made to `staging` and not `master`.

### CODE REVIEW

File: reco_utils/dataset/amazon_reviews.py [hunk 1]
```diff
@@ -233,6 +242,87 @@ def _data_generating(input_file, train_file, valid_file, test_file, min_sequence
...
+                for dt_time in dt_list[:-1]:
+                    dt_str += dt_time + ","
+                if len(cat_str) > 0:
+                    cat_str = cat_str[:-1]
+                if len(mid_str) > 0:
+                    mid_str = mid_str[:-1]
+                if len(dt_str) > 0:
+                    dt_str = dt_str[:-1]
+                if history_clk_num > min_sequence:
+                    fo.write(
```
Comment (loomlike) [inline reco_utils/dataset/amazon_reviews.py:295]: This can be shorten by using a list of strs and `"\t".join(strs)`

File: reco_utils/dataset/amazon_reviews.py [hunk 2]
```diff
@@ -233,6 +242,87 @@ def _data_generating(input_file, train_file, valid_file, test_file, min_sequence
...
+        category = line_split[5]
+
+        if last_tfile == "train":
+            fo = f_train
+        elif last_tfile == "valid":
+            fo = f_valid
+        elif last_tfile == "test":
+            fo = f_test
+        if user_id != last_user_id or tfile == "valid" or tfile == "test":
+            if last_user_id != None:
```
Comment (loomlike) [inline reco_utils/dataset/amazon_reviews.py:277]: To check if none or not, we use `is not None`
> PEP8: Comparisons to singletons like None should always be done with is or is not, never the equality operators.

File: reco_utils/dataset/amazon_reviews.py [hunk 3]
```diff
@@ -233,6 +242,87 @@ def _data_generating(input_file, train_file, valid_file, test_file, min_sequence
...
+                history_clk_num = len(movie_id_list)
+                cat_str = ""
+                mid_str = ""
+                dt_str = ""
+                for c1 in cate_list[:-1]:
```
Comment (loomlike) [inline reco_utils/dataset/amazon_reviews.py:282]: Seems can simply use join.  `",".join(cate_list[:-1])`. Same to the other for-loops.
Btw, what's the last element? Why -1?

File: reco_utils/dataset/amazon_reviews.py [hunk 4]
```diff
@@ -233,6 +242,87 @@ def _data_generating(input_file, train_file, valid_file, test_file, min_sequence
...
+        last_movie_id = movie_id
+        last_category = category
+        last_datetime = date_time
+        last_tfile = tfile
+        if label:
+            movie_id_list.append(movie_id)
+            cate_list.append(category)
+            dt_list.append(date_time)
 
 def _create_item2cate(instance_file):
```
Comment (loomlike) [inline reco_utils/dataset/amazon_reviews.py:327]: What's the reason to use global variable here instead of returning `item2cate`?

File: reco_utils/evaluation/python_evaluation.py [hunk 5]
```diff
@@ -357,7 +357,7 @@ def merge_ranking_true_pred(
         col_item (str): column name for item
         col_rating (str): column name for rating
         col_prediction (str): column name for prediction
-        relevancy_method (str): method for determining relevancy ['top_k', 'by_threshold']
+        relevancy_method (str): method for determining relevancy ['top_k', 'by_threshold', None]
```
Comment (loomlike) [inline reco_utils/evaluation/python_evaluation.py:360]: What's the effect of setting `None`? Can we describe more?

File: reco_utils/evaluation/python_evaluation.py [hunk 6]
```diff
@@ -648,17 +650,21 @@ def get_top_k_items(
         customerID-itemID-rating)
         col_user (str): column name for user
         col_rating (str): column name for rating
-        k (int): number of items for each user
+        k (int or None): number of items for each user; None means that the input has already been
```
Comment (loomlike) [inline reco_utils/evaluation/python_evaluation.py:653]: Is this something related to `relevancy_method=None`?

File: reco_utils/recommender/deeprec/deeprec_utils.py [hunk 7]
```diff
@@ -298,6 +329,10 @@ def create_hparams(flags):
         # dkn
         wordEmb_file=flags["wordEmb_file"] if "wordEmb_file" in flags else None,
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/deeprec_utils.py:330]: We can use `dict.get(key, None)` here

File: reco_utils/recommender/deeprec/deeprec_utils.py [hunk 8]
```diff
@@ -413,6 +448,15 @@ def create_hparams(flags):
         att_fcn_layer_sizes=flags["att_fcn_layer_sizes"]
         if "att_fcn_layer_sizes" in flags
         else None,
+        # nextitnet
+        dilations=flags["dilations"] if "dilations" in flags else None,
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/deeprec_utils.py:452]: Same to the previous comment. `.get(key, None)` will make the codes cleaner.

File: reco_utils/recommender/deeprec/io/dkn_iterator.py [hunk 9]
```diff
@@ -88,27 +106,22 @@ def parser_one_line(self, line):
...
+        userid = cols[1]
+        candidate_news = cols[2]
 
-            elif tokens[0] == "entity":
-                for item in tokens[1].split(","):
-                    candidate_news_entity_index.append(int(item))
-            elif "entity" in tokens[0]:
-                for item in tokens[1].split(","):
-                    click_news_entity_index.append(int(item))
+        for item in self.news_word_index[candidate_news].split(","):
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/dkn_iterator.py:112]: We can use list comprehension, which is more efficient and clean.

File: reco_utils/recommender/deeprec/io/nextitnet_iterator.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+        """Initialize an iterator. Create necessary placeholders for the model.
+        * different from sequential iterator
+        
+        Args:
+            hparams (obj): Global hyper-parameters. Some key settings such as #_feature and #_field are there.
+            graph (obj): the running graph. All created placeholder will be added to this graph.
+            col_spliter (str): column spliter in one line.
+        """
+        self.col_spliter = col_spliter
+        user_vocab, item_vocab, cate_vocab = (
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/nextitnet_iterator.py:33]: Doesn't this create an unnecessary tuple and unwrap back? 🤔 
Also this doesn't make the code shorter.

File: reco_utils/recommender/deeprec/DataModel/ImplicitCF.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        if self.user_idx is None:
+            user_idx = df[[self.col_user]].drop_duplicates().reindex()
+            user_idx[self.col_user + "_idx"] = np.arange(len(user_idx))
+            self.n_users = len(user_idx)
+            self.user_idx = user_idx
+
+            self.user2id = dict(
+                zip(user_idx[self.col_user], user_idx[self.col_user + "_idx"])
+            )
+            self.id2user = {self.user2id[k]: k for k in self.user2id}
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/DataModel/ImplicitCF.py:84]: Reverse table can be simply
`dict(zip(user_idx[self.col_user + "_idx"], user_idx[self.col_user]))`.
This will be faster as well.

File: reco_utils/recommender/deeprec/DataModel/ImplicitCF.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        if self.item_idx is None:
+            item_idx = df[[self.col_item]].drop_duplicates()
+            item_idx[self.col_item + "_idx"] = np.arange(len(item_idx))
+            self.n_items = len(item_idx)
+            self.item_idx = item_idx
+
+            self.item2id = dict(
+                zip(item_idx[self.col_item], item_idx[self.col_item + "_idx"])
+            )
+            self.id2item = {self.item2id[k]: k for k in self.item2id}
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/DataModel/ImplicitCF.py:95]: Same here

File: reco_utils/recommender/deeprec/io/dkn_iterator.py [hunk 13]
```diff
@@ -62,6 +62,24 @@ def __init__(self, hparams, graph, col_spliter=" ", ID_spliter="%"):
             self.click_news_entity_values = tf.placeholder(
                 tf.int64, [None], name="click_news_entity"
             )
+        self.news_word_index = {}
+        self.news_entity_index = {}
+        with tf.gfile.GFile(hparams.news_feature_file, "r") as rd:
+            while True:
+                line = rd.readline()
+                if not line:
+                    break
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/dkn_iterator.py:71]: I think we can simply:
`for line in rd:`

instead of
```
while True:
    line = rd.readline()
    if not line:
        break
```

File: reco_utils/recommender/deeprec/io/dkn_iterator.py [hunk 14]
```diff
@@ -62,6 +62,24 @@ def __init__(self, hparams, graph, col_spliter=" ", ID_spliter="%"):
...
+                newsid, word_index, entity_index = line.strip().split(col_spliter)
+                self.news_word_index[newsid] = word_index
+                self.news_entity_index[newsid] = entity_index
+        self.user_history = {}
+        with tf.gfile.GFile(hparams.user_history_file, "r") as rd:
+            while True:
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/dkn_iterator.py:77]: Same here, `for line in rd:`

File: reco_utils/recommender/deeprec/io/dkn_iterator.py [hunk 15]
```diff
@@ -129,6 +142,8 @@ def load_data_from_file(self, infile):
 
         Returns:
             obj: An iterator that will yields parsed results, in the format of graph feed_dict.
+            List: impression id list
+            Int: size of the data in a batch
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/dkn_iterator.py:146]: I think the codes in this function can be dramatically reduced by using a 2d list containing all the columns. E.g.,
```
candidate_news_index_batch = []
        candidate_news_val_batch = []
        click_news_index_batch = []
        click_news_val_batch = []
        candidate_news_entity_index_batch = []
        click_news_entity_index_batch = []
        label_list = []
        impression_id_list = []

...

(
                    label,
                    candidate_news_index,
                    candidate_news_val,
                    click_news_index,
                    click_news_val,
                    candidate_news_entity_index,
                    click_news_entity_index,
                    impression_id,
                ) = self.parser_one_line(line)

                candidate_news_index_batch.append(candidate_news_index)
                candidate_news_val_batch.append(candidate_news_val)
                click_news_index_batch.append(click_news_index)
                click_news_val_batch.append(click_news_val)
                candidate_news_entity_index_batch.append(candidate_news_entity_index)
                click_news_entity_index_batch.append(click_news_entity_index)
                label_list.append(label)
                impression_id_list.append(impression_id)
```
can be simply:
```
data = [] 
...
    data.append(self.parser_one_line(line))
```

File: reco_utils/recommender/deeprec/io/dkn_iterator.py [hunk 16]
```diff
@@ -187,6 +204,96 @@ def load_data_from_file(self, infile):
...
+                    newsid_list = []
+                    cnt = 0
+
+            if cnt > 0:
+                data_size = cnt
+                while cnt < self.batch_size:
+                    candidate_news_index_batch.append(candidate_news_index_batch[cnt%data_size])
+                    candidate_news_val_batch.append(candidate_news_val_batch[cnt%data_size])
+                    candidate_news_entity_index_batch.append(candidate_news_entity_index_batch[cnt%data_size])
+                    cnt += 1
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/dkn_iterator.py:289]: This function has some logics that are not tested in our unit-tests. It would be nice to have unit tests for this iterator. The only test we have is:

```
iterator = DKNTextIterator(hparams, tf.Graph())
assert iterator is not None
```

File: reco_utils/recommender/deeprec/io/nextitnet_iterator.py [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+            res["item_history"] = item_history_batch_all
+            res["item_cate_history"] = item_cate_history_batch_all
+            res["mask"] = mask
+            res["time"] = time_list_all
+            res["time_diff"] = time_diff_batch
+            res["time_from_first_action"] = time_from_first_action_batch
+            res["time_to_now"] = time_to_now_batch
+
+            # print("label_list_all.shape: ", res["labels"].shape)
+            # print("item_list_all.shape: ", res["items"].shape)
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/nextitnet_iterator.py:217]: Let's clean up the commented-out codes.

File: reco_utils/recommender/deeprec/io/nextitnet_iterator.py [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+            time_diff_batch = np.zeros((instance_cnt, max_seq_length_batch)).astype(
+                "float32"
+            )
+            time_from_first_action_batch = np.zeros(
+                (instance_cnt, max_seq_length_batch)
+            ).astype("float32")
+            time_to_now_batch = np.zeros((instance_cnt, max_seq_length_batch)).astype(
+                "float32"
+            )
+            mask = np.zeros((instance_cnt, max_seq_length_batch)).astype("float32")
```
Comment (loomlike) [inline reco_utils/recommender/deeprec/io/nextitnet_iterator.py:240]: All the `np.zeros(...).astype(..)` should be `np.zeros(..., dtype=...)` to avoid unnecessary array creation.

**Comment by loomlike** [comment]
Ok done for py files. General comments:
1. It would be nice to have better unit-test coverage for the deeprec additions.
2. Some codes could be more simpler, easier to read and maintain w/ some cleanup techniques, like using a list or dict instead of each repetitive variables, etc.

Thank you!!! 🤟

File: README.md [hunk 19]
```diff
@@ -48,40 +48,42 @@ python -m ipykernel install --user --name reco_base --display-name "Python (reco
...
+| Short-term and Long-term preference Integrated Recommender (SLi-Rec)<sup>*</sup> | [Python CPU / Python GPU](examples/00_quick_start/sequential_recsys_amazondataset.ipynb) | Collaborative Filtering | Sequential-based algorithm that aims to capture both long and short-term user preferences using attention mechanism, a time-aware controller and a content-aware controller |
+| Surprise/Singular Value Decomposition (SVD) | [Python CPU](examples/02_model_collaborative_filtering/surprise_svd_deep_dive.ipynb) | Collaborative Filtering | Matrix factorization algorithm for predicting explicit rating feedback in datasets that are not very large |
+| Term Frequency - Inverse Document Frequency (TF-IDF) | [Python CPU](examples/00_quick_start/tfidf_covid.ipynb) | Content-Based Filtering | Simple similarity-based algorithm for content-based recommendations with text datasets |
+| Vowpal Wabbit Family (VW)<sup>*</sup> | [Python CPU (online training)](examples/02_model_content_based_filtering/vowpal_wabbit_deep_dive.ipynb) | Content-Based Filtering | Fast online learning algorithms, great for scenarios where user features / context are constantly changing |
```
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply README.md:84]: changed

File: README.md [hunk 20]
```diff
@@ -3,11 +3,11 @@
...
+- [Model](examples/00_quick_start): Building models using various classical and deep learning recommender algorithms such as Alternating Least Squares ([ALS](https://spark.apache.org/docs/latest/api/python/_modules/pyspark/ml/recommendation.html#ALS)) or eXtreme Deep Factorization Machines ([xDeepFM](https://arxiv.org/abs/1803.05170)).
+- [Evaluate](examples/03_evaluate): Evaluating algorithms with offline metrics
+- [Model Select and Optimize](examples/04_model_select_and_optimize): Tuning and optimizing hyperparameters for recommender models
+- [Operationalize](examples/05_operationalize): Operationalizing models in a production environment on Azure
 
 Several utilities are provided in [reco_utils](reco_utils) to support common tasks such as loading datasets in the format expected by different algorithms, evaluating model outputs, and splitting training/test data. Implementations of several state-of-the-art algorithms are included for self-study and customization in your own applications. See the [reco_utils documentation](https://readthedocs.org/projects/microsoft-recommenders/).
```
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply README.md:13]: fixed all those I could find

File: reco_utils/dataset/amazon_reviews.py (hunk 2 above)
Comment (miguelgfierro) [reply reco_utils/dataset/amazon_reviews.py:277]: good catch!

File: reco_utils/recommender/deeprec/deeprec_utils.py (hunk 7 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/deeprec_utils.py:330]: I completely agree with you, this is more pythoninc, in fact, we could do directly `dict.get(key)` because the default is alternative is `None`. However, changing this piece of code would break a lot of the code on deeprec and I guess probably there will be a lot of refactoring needed. 

I would leave it as it is, unless @Leavingseason thinks we should change it

File: reco_utils/recommender/deeprec/io/nextitnet_iterator.py (hunk 17 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/nextitnet_iterator.py:217]: done

File: reco_utils/recommender/deeprec/io/nextitnet_iterator.py (hunk 18 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/nextitnet_iterator.py:240]: good catch, changed

File: reco_utils/dataset/amazon_reviews.py [hunk 21]
```diff
@@ -233,6 +242,87 @@ def _data_generating(input_file, train_file, valid_file, test_file, min_sequence
             cate_list.append(category)
             dt_list.append(date_time)
 
+def _data_generating_no_history_expanding(input_file, train_file, valid_file, test_file, min_sequence=1):
```
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply reco_utils/dataset/amazon_reviews.py:245]: added as an issue to solve on a different PR https://github.com/microsoft/recommenders/issues/1138

**Comment by miguelgfierro** [comment]
hey @loomlike, I started to fix some of the issues. If you have time, feel free to push some of those. Tao suggested closing the PR this week

File: reco_utils/dataset/amazon_reviews.py (hunk 4 above)
Comment (miguelgfierro) [reply reco_utils/dataset/amazon_reviews.py:333]: mmm, I don't see a justification to use a global here, there are multiple problems in this function, created an issue to fix them on a different pr https://github.com/microsoft/recommenders/issues/1141

File: reco_utils/dataset/amazon_reviews.py (hunk 3 above)
Comment (miguelgfierro) [reply reco_utils/dataset/amazon_reviews.py:287]: https://github.com/microsoft/recommenders/issues/1141

File: reco_utils/dataset/amazon_reviews.py (hunk 1 above)
Comment (miguelgfierro) [reply reco_utils/dataset/amazon_reviews.py:300]: https://github.com/microsoft/recommenders/issues/1141

File: reco_utils/evaluation/python_evaluation.py (hunk 5 above)
Comment (miguelgfierro) [reply reco_utils/evaluation/python_evaluation.py:360]: added comments

File: reco_utils/evaluation/python_evaluation.py (hunk 6 above)
Comment (miguelgfierro) [reply reco_utils/evaluation/python_evaluation.py:653]: added comments

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 14 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:77]: fixed, good catch!

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 13 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:71]: fixed, good catch!

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 15 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:141]: I'm not sure about this tbh, yes, it is cleaner, but at the same time the way it is now, it shows clearly all the components. This looks like a programming choice.

I think both options are equally valid, so I would ask @leavingseason if he wants to spend time in changing this

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 16 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:293]: I think we can do this on another PR

File: reco_utils/recommender/deeprec/io/nextitnet_iterator.py (hunk 10 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/nextitnet_iterator.py:33]: fixed

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 9 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:107]: this case has two instructions:
```
        for item in self.news_word_index[candidate_news].split(","):
            candidate_news_index.append(int(item))
            candidate_news_val.append(float(1))
```
and there are several others. 

One can argue that showing it the way it is, it might be clearer.

I think both options are valid, so I would ask @Leavingseason if he wants to spend time in changing this

File: reco_utils/recommender/deeprec/DataModel/ImplicitCF.py (hunk 11 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/DataModel/ImplicitCF.py:84]: done

File: reco_utils/recommender/deeprec/DataModel/ImplicitCF.py (hunk 12 above)
Comment (miguelgfierro) [reply reco_utils/recommender/deeprec/DataModel/ImplicitCF.py:95]: done

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 15 above)
Comment (loomlike) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:141]: yes I agree that that is a programming choice. Then I'm okay w/ the way it is now.

File: reco_utils/recommender/deeprec/deeprec_utils.py (hunk 7 above)
Comment (loomlike) [reply reco_utils/recommender/deeprec/deeprec_utils.py:330]: Yes, let's leave this up to the original author.

File: reco_utils/recommender/deeprec/io/dkn_iterator.py (hunk 9 above)
Comment (loomlike) [reply reco_utils/recommender/deeprec/io/dkn_iterator.py:107]: Got it. Let's revisit this later if we'll have a performance issue.
